File: JAVA/kastenlauf/kastenlauf.java
/*
Rating: ~ 3.9 / 10
Link: https://open.kattis.com/problems/kastenlauf
*/

import java.io.*;
import java.util.*;

public class kastenlauf {
   public static void main(String[] args) {
      Scanner scan = new Scanner(System.in);
      int cases = Integer.parseInt(scan.nextLine());
      for (int x = 0; x < cases; x++) {
         int stops = Integer.parseInt(scan.nextLine());
         Coordinates[] bars = new Coordinates[stops];
         Coordinates start = new Coordinates(scan.nextInt(), scan.nextInt());
         scan.nextLine();
         for (int y = 0; y < stops; y++) {
            bars[y] = new Coordinates(scan.nextInt(), scan.nextInt());
            scan.nextLine();
         }
         Coordinates end = new Coordinates(scan.nextInt(), scan.nextInt());
         scan.nextLine();
         Stack<Coordinates> track = new Stack<Coordinates>();
         boolean[] visited = new boolean[bars.length];
         track.push(start);
         boolean found = DFS(track, end, bars, visited);
         if (found) {
            System.out.println("happy");
         }
         else {
            System.out.println("sad");
         }
      }
   }

   public static boolean DFS(Stack<Coordinates> track, Coordinates end, Coordinates[] bars, boolean[] visited) {
      Coordinates current = track.pop();
      int distance = Math.abs(current.x-end.x)+Math.abs(current.y-end.y);
      if (distance <= 1000) {
         return true;
      }
      boolean found = false;
      for (int x = 0; x < bars.length; x++) {
         if (visited[x]) {
            continue;
         }
         if (found) {
            break;
         }
         int checkd = Math.abs(current.x-bars[x].x)+Math.abs(current.y-bars[x].y);
         if (checkd <= 1000) {
            track.push(current);
            track.push(bars[x]);
            visited[x] = true;
            found = DFS(track, end, bars, visited);
         }
      }
      return found;
   }
}

class Coordinates {
   int x = -5;
   int y = -5;

   Coordinates(int x, int y) {
      this.x = x;
      this.y = y;
   }
}

File: PYTHON/kastenlauf/kastenlauf.py
from sys import stdin, stdout
from collections import deque

# Rating: ~ 4.0 / 10
# Link: https://open.kattis.com/problems/kastenlauf

def dfs(start, end, stores):
  stack = [start]
  visited = set()

  while stack:
    x, y = stack.pop()

    if abs(x - end[0]) + abs(y - end[1]) <= 1000:
      return 'happy'

    for store in stores:
      if store not in visited and abs(x - store[0]) + abs(y - store[1]) <= 1000:
        stack.append(store)
        visited.add(store)

  return 'sad'


def main():
  cases = int(stdin.readline().strip())

  for x in range(cases):
    start = None
    stores = set()
    end = None
    numS = int(stdin.readline().strip())

    start = tuple(int(a) for a in stdin.readline().split())
    for y in range(numS):
      x, y = map(int, stdin.readline().split())
      stores.add((x, y))
    end = tuple(int(a) for a in stdin.readline().split())

    print(dfs(start, end, stores))


if __name__ == "__main__":
  main()
